### ISSUE TITLE
Warn if make iterator out of iterator in `for`.

### ISSUE THREAD

**Issue by bmusin** [rust-lang-nursery/rust-clippy#2369]
Hi.
```
fn main() {
    // for m in (1..10).take(5) is enough!
    for m in (1..10).take(5).collect::<Vec<i32>>() {
        println!("{}", m);
    }
}
```
`.take(5)` gives iterator that is usable in  `for` loop.
`.collect::<Vec<i32>>()` creates `Vec` from `.take` iterator and it(`Vec`) has iterator behaviour too.

What about creating lint, that warns, because code creates iterator out of iterator unnecessarily?
Thanks.

**Comment by camsteffen** [comment]
This should trigger `needless_collect`. We just have to be careful there is not something like `map(...)` with side effects.
